tests/integration/__init__.py

```diff
@@ -8,16 +9,116 @@ class IntegrationTestBase(TestBase):
+        parser extraction."""
+        sub_path = self.subdir.split("/")[-1]
+        expected = self.references[self.version][sub_path][self.property]
+        reference_index = self.test_config.pop("reference_index", None)
+        if reference_index is not None:
+            return expected[reference_index]
+        return expected
 
-    @property
-    def workDir(self):
-        return os.path.join(self.rootDir, self.getManifest()["workDir"])
 
-    @property
-    def stdoutFile(self):
-        return os.path.join(self.rootDir, self.getManifest()["stdoutFile"])
+def get_test_config(test_config: Union[str, dict]) -> dict:
+    """If specified at the individual test level, construct a granular test
+    configuration, otherwise provide a default test configuration from the
+    manifest.yaml.
+
+    Note:
+        See also the test_config_schema in the manifest.yaml file and
+        example test_configs for usage. Test config keys that are not recognized
+        as keyword args to comparison methods must be popped during processing
+        or the comparison method will throw an unrecognized argument error.
+    """
+    places = 2
+    if isinstance(test_config, str):
+        return {
+            "property": test_config,
+            "places": places,
+            **TestBase.manifest["tests"][test_config],
+        }
+    return {
+        **TestBase.manifest["tests"][test_config["name"]],
+        "property": test_config.pop("name"),
+        "places": places,
+        **test_config,
+    }
+
+
+def create_test(params: dict):
+    """Function factory that generates a uniquely named TestCase method
+    based on the required input parameters:
+
+    Args:
+        params(dict): {
+            "application": application name,
+            "version": version string,
+            "subdir": subdir to fixture,
+            "filename": test filename,
+            "property": test property,
+            "test_config": test config from manifest,
+        }
+
+    Returns:
+        func(function): a unittest.TestCase.test_method
+
+    """
+    clean = lambda s: s.replace(".", "_").replace("/", "_").replace("-", "_")
+    test_names = ["test"]
+    for k, v in params.items():
+        if isinstance(v, str):
+            if k == "version":
+                test_names.append(clean(v.replace(".", "")))
+            else:
+                test_names.append(clean(v))
+    def do_test(self):
+        for k, v in params.items():
+            setattr(self, k, v)
+        work_dir = self.work_dir(self.version, self.subdir)
+        # TODO: os.path.join(workdir, self.filename) should not be necessary
+        parser = self.parser(
+            work_dir=work_dir, stdout_file=os.path.join(work_dir, self.filename)
+        )
+        actual = getattr(parser, self.property)()
+        actual_index = self.test_config.pop("actual_index", None)
+        if actual_index is not None:
+            actual = actual[actual_index]
```

Comment:
I suggest splitting the above code into 2 parts:

1. get parser and actual value
2. get comparison operation and expected value 

The logic dealing with `actual_index` etc would be dealt with in 1, and not in the "thick" of a nested `do_test` function. The latter can be a function of the TestBase class also, as well as 1 and 2 above.

Reply:
Moved logic into a `get_actual_value` method to mimic the `get_reference_value` method, cleans up the `do_test` function quite a bit.